_posts/2024-01-10-semantics-of-regular-expressions.markdown

```diff
@@ -0,0 +1,386 @@
+
+```
+lemma DenotationalIsAlgebraHomomorphism<A(!new)>()
+  ensures IsAlgebraHomomorphism<A>(Denotational)
+{
+  forall e ensures IsAlgebraHomomorphismPointwise<A>(Denotational, e) {
+    BisimilarityIsReflexive<A>(Denotational(e));
+  }
+}
+```
+
+## Operational Semantics
+
+In this section, we provide an alternative perspective on the semantics of regular expressions. We equip the set of regular expressions with a coalgebraic structure, formalise its *operational* semantics as a function from regular expressions to formal languages, and prove that the latter is a coalgebra homomorphism.
+
+### A Coalgebra of Regular Expressions
+
+In [An Algebra of Formal Languages](#an-algebra-of-formal-languagesn Algebra of Formal Languages) we equipped the set of formal languages with an algebraic structure that resembles the one of regular expressions. Now, we are aiming for the reverse: we would like to equip the set of regular expressions with a coalgebraic structure that resembles the one of formal languages. More concretely, we would like to turn the set of regular expressions into a deterministic automaton (without initial state) in which a state `e` is i) accepting iff `Eps(e) == true` and ii) transitions to a state `Delta(e)(a)` if given the input `a: A`. Note how our definitions resemble the Brzozowski derivatives we previously encountered:
+
+```
+ghost function Eps<A>(e: Exp): bool {
+  match e 
+    case Zero => false
+    case One => true
+    case Char(a) => false
+    case Plus(e1, e2) => Eps(e1) || Eps(e2)
+    case Comp(e1, e2) => Eps(e1) && Eps(e2)
+    case Star(e1) => true
+}
+
+ghost function Delta<A(!new)>(e: Exp): A -> Exp {
+  (a: A) => 
+    match e 
+      case Zero => Zero
+      case One => Zero
+      case Char(b) => if a == b then One else Zero
+      case Plus(e1, e2) => Plus(Delta(e1)(a), Delta(e2)(a))
+      case Comp(e1, e2) => Plus(Comp(Delta(e1)(a), e2), Comp(if Eps(e1) then One else Zero, Delta(e2)(a)))
+      case Star(e1) => Comp(Delta(e1)(a), Star(e1))
+}
+```
+
+### Operational Semantics as Induced Morphism
+
+The operational semantics of regular expressions can now be defined via coinduction, as a function `Operational: Exp -> Lang`, by making use of the coalgebraic structure on expressions we have just defined:
+
+```
+function Operational<A(==)>(e: Exp): Lang {
+  Alpha(Eps(e), (a: A) => Operational(Delta(e)(a)))
+}
+```
+
+### Operational Semantics as Coalgebra Homomorphism
+
+In [Denotational Semantics as Algebra Homomorphism](#denotational-semantics-as-algebra-homomorphism) we defined algebra homomorphisms as functions `f: Exp -> Lang` that commute with the algebraic structures of regular expressions and formal languages, respectively. Analogously, let us now call a function `f` of the same type a *coalgebra homomorphism*, if it commutes with the *coalgebraic* structures of regular expressions and formal languages, respectively:
```

Comment:
The link is broken in the Github rendering. Not sure how it'll render on the blog.

Reply:
The link works